File: app/lib/core/decode/beat_sync.dart
// Beat-synchronous chord decode: one chord per beat interval (mode of the
// model's per-frame class ids in that interval), consecutive equal labels
// merged. Anchoring chord changes to real beats removes sub-beat flicker
// (#5) and keeps the grid in sync with the audio (#4). Falls back to
// voteDecode (see on_device_analyzer) when there are no beats.
import '../inference/pcm_runner.dart';
import '../model_registry.dart';
import '../models.dart';
import 'vote_decode.dart' show fallbackFrameDur;

List<Chord> beatSyncChords(
  List<FrameResult> frames,
  List<double> beatTimes,
  ModelSpec spec,
) {
  if (frames.isEmpty || beatTimes.isEmpty) return const [];
  final labels = spec.labels;
  if (labels == null) {
    throw ArgumentError('${spec.name} has no labels in its manifest entry');
  }

  final nIndex = labels.indexOf('N'); // -1 if absent

  final frameDur =
      frames.length > 1 ? frames[1].time - frames[0].time : fallbackFrameDur;
  final songEnd = frames.last.time + frameDur;

  // Interval boundaries: optional lead-in from 0, the beats, then song end.
  final bounds = <double>[];
  if (beatTimes.first > 0) bounds.add(0.0);
  bounds.addAll(beatTimes);
  final end = songEnd > bounds.last ? songEnd : bounds.last + frameDur;
  bounds.add(end);

  // Walk frames once (both frames and bounds are ascending in time).
  final segments = <Chord>[];
  var fi = 0;
  var prevClass = -1;
  for (var b = 0; b + 1 < bounds.length; b++) {
    final lo = bounds[b], hi = bounds[b + 1];
    final counts = <int, int>{};
    final confSum = <int, double>{};
    while (fi < frames.length && frames[fi].time < hi) {
      if (frames[fi].time >= lo) {
        final c = frames[fi].classId;
        counts[c] = (counts[c] ?? 0) + 1;
        confSum[c] = (confSum[c] ?? 0) + frames[fi].confidence;
      }
      fi++;
    }

    int winner;
    double conf;
    if (counts.isEmpty) {
      // Empty interval: inherit the previous chord (or N for the first).
      winner = prevClass >= 0 ? prevClass : (nIndex >= 0 ? nIndex : 0);
      conf = 0.0;
    } else {
      winner = counts.keys.first;
      for (final c in counts.keys) {
        final better = counts[c]! > counts[winner]! ||
            (counts[c] == counts[winner] && confSum[c]! > confSum[winner]!);
        if (better) winner = c;
      }
      conf = confSum[winner]! / counts[winner]!;
    }
    prevClass = winner;

    // Merge into the previous segment if the label matches.
    if (segments.isNotEmpty && segments.last.chord == labels[winner]) {
      final prev = segments.removeLast();
      final prevDur = prev.end - prev.start;
      final curDur = hi - lo;
      segments.add(Chord.fromJson({
        'chord': prev.chord,
        'start': prev.start,
        'end': hi,
        'confidence':
            (prev.confidence * prevDur + conf * curDur) / (prevDur + curDur),
      }));
    } else {
      segments.add(Chord.fromJson({
        'chord': labels[winner], 'start': lo, 'end': hi, 'confidence': conf,
      }));
    }
  }
  return segments;
}

File: server/app/infrastructure/analysis/beat_sync.py
from collections import defaultdict
from .chords import Chord
from .onnx_infer import Frame
from .manifest import ModelSpec
from .vote_decode import majority_filter, merge_short_chords, FALLBACK_FRAME_DUR


def beat_sync_chords(frames: list[Frame], beat_times: list[float], spec: ModelSpec,
                     beat_smoothing_kernel: int = 3, min_chord_dur: float = 0.0) -> list[Chord]:
    if not frames or not beat_times:
        return []
    labels = spec.labels
    n_index = labels.index("N") if "N" in labels else -1
    frame_dur = frames[1].time - frames[0].time if len(frames) > 1 else FALLBACK_FRAME_DUR
    song_end = frames[-1].time + frame_dur

    bounds: list[float] = []
    if beat_times[0] > 0:
        bounds.append(0.0)
    bounds.extend(beat_times)
    end = song_end if song_end > bounds[-1] else bounds[-1] + frame_dur
    bounds.append(end)

    winners: list[int] = []
    confs: list[float] = []
    fi = 0
    prev_class = -1
    for b in range(len(bounds) - 1):
        lo, hi = bounds[b], bounds[b + 1]
        counts: dict[int, int] = defaultdict(int)
        conf_sum: dict[int, float] = defaultdict(float)
        while fi < len(frames) and frames[fi].time < hi:
            if frames[fi].time >= lo:
                c = frames[fi].class_id
                counts[c] += 1
                conf_sum[c] += frames[fi].confidence
            fi += 1
        if not counts:
            winner = prev_class if prev_class >= 0 else (n_index if n_index >= 0 else 0)
            conf = 0.0
        else:
            winner = next(iter(counts))
            for c in counts:
                if counts[c] > counts[winner] or (counts[c] == counts[winner] and conf_sum[c] > conf_sum[winner]):
                    winner = c
            conf = conf_sum[winner] / counts[winner]
        prev_class = winner
        winners.append(winner)
        confs.append(conf)

    smoothed = majority_filter(winners, beat_smoothing_kernel)

    segments: list[Chord] = []
    for b in range(len(smoothed)):
        lo, hi = bounds[b], bounds[b + 1]
        label = labels[smoothed[b]]
        conf = confs[b]
        if segments and segments[-1].chord == label:
            prev = segments.pop()
            prev_dur = prev.end - prev.start
            cur_dur = hi - lo
            segments.append(Chord(prev.chord, prev.start, hi,
                                  (prev.confidence * prev_dur + conf * cur_dur) / (prev_dur + cur_dur)))
        else:
            segments.append(Chord(label, lo, hi, conf))

    return merge_short_chords(segments, min_chord_dur) if min_chord_dur > 0 else segments
